### ISSUE TITLE
Support dynamic sources when specifying policy routing rules

### ISSUE THREAD

**Issue by poscat0x04** [systemd/systemd#25971]
### Component

systemd-networkd

### Is your feature request related to a problem? Please describe

I'm using systemd-networkd to manage a machine that has two interfaces with different addresses that can potentially change (they are acquired dynamically using router solicitation and ppp). One of the interfaces is the default interface for traffic, but I'd like the machine to avoid asymmetric routing, that is, response to traffic coming from the other interface should egress from the same interface. This means I'll have to add routes or policy routing rules with the source matching the IP addresses of my interfaces. But since the addresses are dynamic (and can even expire), I have no way of doing that using systemd-networkd currently.

### Describe the solution you'd like

I'd like systemd-networkd to support routes whose sources are dynamic. The solution would look something like this:

A special value, for example `self` is added to the possible values of the `Source` attribute under `[Route]`, which, when specified, for each address that interface has, add a route with the `Source` attribute set to that address. The routes should also change accordingly when the IP address on the interfaces change.

Another option would be to use policy routing rules, but since rules cannot have the same priority (whereas different routes can have the same metric), the solution could be a bit more complicated.

### Describe alternatives you've considered

An alternative would be to write a daemon that monitors the interface and adds/changes the routes when the IP addresses change. But then there will be quite a bit of duplicated effort.

### The systemd version you checked that didn't have the feature you are asking for

251.7

**Comment by TriMoon** [comment]
> A special value, for example `self` is added to the possible values of the `Source` attribute under `[Route]`

Uhmm.... Unless i misunderstood you...
See: Section `[ROUTE] SECTION OPTIONS` in `systemd.network(5)`:
> Source=
> The source prefix of the route. Possibly followed by a slash and the prefix length.
 > <b><i>If omitted, a full-length host route is assumed.</i></b>

Which means your option-value of `self` is already implemented by **not specifying** the option `Source=` at all. :woman_shrugging:

**Comment by poscat0x04** [comment]
Well, I misunderstood what `Source=` means. It turns out `Source=` corresponds to `src` which only affects address selection. The `From=` attribute under `[ROUTINGPOLICYRULE]` is what I meant then.

**Comment by TriMoon** [comment]
Maybe use a combination with `IncomingInterface=`  :thinking:

**Comment by poscat0x04** [comment]
You mean `OutgoingInterface=`? Cause this issue is concerned with traffic originated from the machine. The description says it is "only available for packets originating from local sockets that are bound to a device". I'm not sure what that means but it seems that it does not apply to all packets?

**Comment by poscat0x04** [comment]
The thing is in policy-based routing there's no concept of connections. Policy-based routing rules are applied on individual packets, which means for the `OutgoingInterface=` and `IncomingInterface=` are reversed for responding packets.

I'll give an example for more clarity:
Suppose I have a machine with two interfaces `if1` and `if2`, `if1` has the address `1.1.1.1/32` and `if2` has the address `1.0.0.1/32`, both are connected to the internet. I've configured `if1` as the default interface, in other words, there's this route

```
0.0.0.0/0 via dev if1
```

In the default routing table. There's also an application that's listening on the tcp port `443` on `1.0.0.1`. Upon receiving a connection, the kernel responds with syn-ack packet. According to the tcp specification, the source IP address and the destination IP address should be reversed, hence the source IP address of the syn-ack packet will be set to `1.0.0.1`. **But the kernel does not look at source IP addresses while making routing decisions**. It only sees that the destination IP address matches the route mentioned above and decides to send it through `if2`. Note that this has nothing to do with IP forwarding, this can happen even if `net/ipv4/ip_forward` is turned off, since the kernel **need not to make routing decisions on incoming packets**.

Anyways I eventually solved it using nftables and conntrack. Though I'm not sure if I should leave this issue open.
